README.md

```diff
@@ -1,28 +1,411 @@
+- **Event Versioning**: Support for tracking versions of events for schema evolution
+- **Actor Attribution**: Store who performed each action (users, system processes, etc)
+- **Event Replay**: Rebuild resource state by replaying events
+- **Version-specific Replay Routing**: Route events to different actions based on their version
+- **Customizable Metadata**: Attach arbitrary metadata to events
+
+## Installation
+
+Add `ash_events` to your dependencies in `mix.exs`:
+
+```elixir
+def deps do
+  [
+    {:ash_events, "~> 0.1.0"}
+    # ... other deps
+  ]
+end
+```
+
+## Usage
+
+### 1. Create an Event Log Resource
+
+First, define a resource that will store your events:
+
+```elixir
+defmodule MyApp.Events.Event do
+  use Ash.Resource,
+    extensions: [AshEvents.EventLog]
+
+  event_log do
+    # Module that implements clear_records! callback
+    clear_records_for_replay MyApp.Events.ClearAllRecords
+
+    # Optional, defaults to :uuid
+    record_id_type :uuid
+
+    # Store primary key of actors running the actions
+    persist_actor_primary_key :user_id, MyApp.Accounts.User
+    persist_actor_primary_key :system_actor, MyApp.SystemActor, attribute: :string
+  end
+
+  # Optional: Configure replay overrides for version handling
+  replay_overrides do
+    replay_override MyApp.Accounts.User, :create do
+      versions [1]
+      route_to MyApp.Accounts.User, :old_create_v1
+    end
+  end
+  .....
+end
+```
+
+### 2. Define a Clear Records Implementation
+
+Implement the module that will clear records before replay:
+
+```elixir
+defmodule MyApp.Events.ClearAllRecords do
+  use AshEvents.ClearRecordsForReplay
+
+  @impl true
+  def clear_records!(opts) do
+    # Logic to clear all relevant records for all resources with event tracking
+    # enabled through the event log resource.
+    :ok
+  end
+end
+```
+
+### 3. Enable Event Logging on Resources
+
+Add the `AshEvents.Events` extension to resources you want to track:
+
+```elixir
+defmodule MyApp.Accounts.User do
+  use Ash.Resource,
+    extensions: [AshEvents.Events]
+
+  events do
+    # Specify your event log resource
+    event_log MyApp.Events.Event
+
+    # Optionally ignore certain actions. This is mainly used for actions
+    # that are kept around for supporting previous event versions, and
+    # are configured as replay_overrrides in the event log (see above).
+    ignore_actions [:old_create_v1]
+
+    # Optionally specify version numbers for actions
+    current_action_versions create: 2, update: 3, destroy: 2
+  end
+
+  # Rest of your resource definition...
+  attributes do
+    # ...
+  end
+
+  actions do
+    # ...
+  end
+end
+```
+
+### 4. Track Metadata with Actions
+
+When performing actions, you can include any metadata by adding the `ash_events_metadata` argument:
+
+```elixir
+User
+|> Ash.Changeset.for_create(:create, %{
+  name: "Jane Doe",
+  email: "jane@example.com",
+  ash_events_metadata: %{
```

Comment:
I think we need to either use `context` for this, or use a private argument that is set via `private_arguments: %{ash_events_metadata: ...}`. Probably context. The reason is that if you try to put this up into an API we'd be exposing an arbitrary input over that interface.